### PR TITLE
New features proposal

### PR DESCRIPTION
**Pull request by bobopinna**
Hi,
I'm using your enrolment plugin and I found some missing features.
I worked on them and I think they could be useful to other users.

The first thing I've added is the method instance deletion from course enrolment methods. I see that someone ask for it in moodle plugin repository. I've just added a missing function in lib.php and changed a capability in db/access.php.

This is the second change I made:
I use user profile fields to manage automatic membership in a moodle forum with separated groups. 
Autoenrol worked fine for the first year but I need that group membership changes for users that has changed user profile datas. 
For example students that pass from the first year to the second year or students that change their academic degree. I hope you can understand what I try to explain because I think there are several differences from english/US and italian university.

At last I've implemented support to other profile fields and custom profile fields.

HTH,
Roberto.

### CODE REVIEW

File: enrol/autoenrol/db/upgrade.php [hunk 1]
```diff
@@ -41,45 +41,68 @@ function xmldb_enrol_autoenrol_upgrade($oldversion) {
...
+        $fields[] = 'auth';
+        $fields[] = 'department';
+        $fields[] = 'institution';
+        $fields[] = 'lang';
+        $fields[] = 'email';
+
+        $instances = $DB->get_records('enrol', array('enrol' => 'autoenrol'));
+
+        foreach ($instances as $instance) {
+            if (isset($instance->customint2)) {
```
Comment (markward) [inline enrol/autoenrol/db/upgrade.php:96]: Hi @bobopinna 

Does customint2 remain populated after this update process? Is there any reason not to clean up the values stored there?

Thanks

Mark

File: enrol/autoenrol/lib.php [hunk 2]
```diff
@@ -54,7 +56,7 @@ public function roles_protected() {
      * @return bool
      */
     public function allow_unenrol(stdClass $instance) {
-        // Users with unenrol cap may unenrol other users manually - requires enrol/autoenrol:unenrol.
+        // Users with onenrol cap may unenrol other users manually - requires enrol/autoenrol:unenrol.
```
Comment (markward) [inline enrol/autoenrol/lib.php:59]: Hi @bobopinna 

Was this change deliberate? I'm just wondering whether this really does mean "on enrol" now?

File: enrol/autoenrol/lib.php [hunk 3]
```diff
@@ -155,50 +161,84 @@ public function enrol_allowed($USER, stdClass $instance) {
...
+                $profileattribute = $instance->customchar3;
+            } else if (isset($instance->customint2)) {
+                $oldfields = array(0 => '', 1 => 'auth', 2 => 'department', 3 => 'institution', 4 => 'lang', 5 => 'email');
+                $profileattribute = $oldfields[$instance->customint2];
             }
 
-            $type = array(1 => $USER->auth, $USER->department, $USER->institution, $USER->lang, $USER->email);
+            $standardfields = array('auth', 'lang', 'department', 'institution', 'address', 'city');
+            if (in_array($profileattribute, $standardfields)) {
+                if (!isset($user->auth)) {
```
Comment (markward) [inline enrol/autoenrol/lib.php:200]: Hi @bobopinna 

It looks like we'd be able to take care of all this through the API below if we were using Moodle 3.1 as a minimum?

https://docs.moodle.org/dev/User-related_APIs#User_fields_definition

What do you think?

**Review by markward — COMMENTED** [review]
Hi @bobopinna 

Thanks for submitting this pull request, it looks like a very good starting point for 2.0.

If left a few items for review, but these are mainly a few quick questions rather than significant changes. 

Cheers!

Mark

File: enrol/autoenrol/db/upgrade.php (hunk 1 above)
Comment (bobopinna) [reply enrol/autoenrol/db/upgrade.php:96]: Sure, it can be cleaned. I forgot it.

File: enrol/autoenrol/lib.php (hunk 2 above)
Comment (bobopinna) [reply enrol/autoenrol/lib.php:59]: No, I don't want to change it. May be it was changed by type checker.
